store/store.go

```diff
@@ -0,0 +1,415 @@
+	heightsPath = blocksPath + "heights/"
+
+	defaultDirPerm = 0o755
+)
+
+var ErrNotFound = errors.New("eds not found in store")
+
+// Store is a storage for EDS files. It persists EDS files on disk in form of Q1Q4 files or ODS
+// files. It provides methods to put, get and remove EDS files. It has two caches: recent eds cache
+// and availability cache. Recent eds cache is used to cache recent blocks. Availability cache is
+// used to cache blocks that are accessed by sample requests. Store is thread-safe.
+type Store struct {
+	// basepath is the root directory of the store
+	basepath string
+	// cache is used to cache recent blocks and blocks that are accessed frequently
+	cache *cache.DoubleCache
+	// stripedLocks is used to synchronize parallel operations
+	stripLock *striplock
+	metrics   *metrics
+}
+
+// NewStore creates a new EDS Store under the given basepath and datastore.
+func NewStore(params *Parameters, basePath string) (*Store, error) {
+	if err := params.Validate(); err != nil {
+		return nil, err
+	}
+
+	// Ensure the blocks folder exists or is created.
+	blocksFolderPath := basePath + blocksPath
+	if err := ensureFolder(blocksFolderPath); err != nil {
+		log.Errorf("Failed to ensure the existence of the blocks folder at '%s': %s", blocksFolderPath, err)
+		return nil, fmt.Errorf("ensure blocks folder '%s': %w", blocksFolderPath, err)
+	}
+
+	// Ensure the heights folder exists or is created.
+	heightsFolderPath := basePath + heightsPath
+	if err := ensureFolder(heightsFolderPath); err != nil {
+		log.Errorf("Failed to ensure the existence of the heights folder at '%s': %s", heightsFolderPath, err)
+		return nil, fmt.Errorf("ensure heights folder '%s': %w", heightsFolderPath, err)
+	}
+
+	err := createEmptyFile(basePath)
+	if err != nil {
+		return nil, fmt.Errorf("creating empty file: %w", err)
+	}
+
+	recentEDSCache, err := cache.NewAccessorCache("recent", params.RecentBlocksCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create recent eds cache: %w", err)
+	}
+
+	availabilityCache, err := cache.NewAccessorCache("blockstore", params.AvailabilityCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create availability cache: %w", err)
+	}
+
+	store := &Store{
+		basepath:  basePath,
+		cache:     cache.NewDoubleCache(recentEDSCache, availabilityCache),
+		stripLock: newStripLock(1024),
+	}
+	return store, nil
+}
+
+func (s *Store) Close() error {
+	return s.metrics.close()
+}
+
+func (s *Store) Put(
+	ctx context.Context,
+	datahash share.DataHash,
+	height uint64,
+	square *rsmt2d.ExtendedDataSquare,
+) (eds.AccessorStreamer, error) {
+	tNow := time.Now()
+	lock := s.stripLock.byDatahashAndHeight(datahash, height)
+	lock.lock()
+	defer lock.unlock()
+
+	if datahash.IsEmptyRoot() {
+		err := s.addEmptyHeight(height)
+		return emptyAccessor, err
+	}
+
+	// short circuit if file exists
+	if has, _ := s.hasByHeight(height); has {
+		s.metrics.observePutExist(ctx)
+		return s.getByHeight(height)
+	}
```

Comment:
optional optimization: 
we could do `getByHeight` here, and if it exists, return directly. In case of an error create a file. Otherwise we do duplicate work across both calls